### PR TITLE
Use explicit destinations in codegen to avoid uselessly jumping around.

### PR DESCRIPTION
**Pull request by sjrd**
~~CI for now. This needs some cleanup and better tests that the bytecode is indeed improved in a number of situations. (notably with pattern matching)~~

Previously, the codegen's main method `genLoad` always generated code that loaded the value on the stack before continuing. There were a number of situations where `genLoad` would be directly followed by unconditional jumps to instructions performing more jumps, returns and throws. This generated more spurious jumps than necessary, along with artifact dead code.

We solve these limitations by introducing `LoadDestination`s that specify the destination of a loaded value:

* `FallThrough`: as previously, load the value on the stack and continue.
* `Jump(label)`: load the value on the stack and jump to the given label.
* `Return`: return the value from the enclosing method.
* `Throw`: throw the value.

We generalize `genLoad` as `genLoadTo`, taking a specific destination for the loaded value. `genLoadTo` can "push down" its destination into all control flow structures (except `Try`s, because of their cleanups). With that, when we get to the end of what amounts to "basic blocks", we know exactly the ultimate destination of the loaded value. We can therefore directly jump, return or throw to the final destination.

This produces less bytecode, notably because fewer labels are necessary. For example, the method:
```scala
  def abs(x: Int): Int = if x < 0 then -x else x
```
previously generated bytecode like
```
  ILOAD 1
  ICONST_0
  IF_ICMPGE Label(1)
  ILOAD 1
  INEG
  GOTO Label(2)
  Label(1):
  ILOAD 1
  Label(2):
  IRETURN
```
Now, instead of jumping to `Label(2)`, we directly perform an `IRETURN`:
```
  ILOAD 1
  ICONST_0
  IF_ICMPGE Label(1)
  ILOAD 1
  INEG
  IRETURN
  Label(1):
  ILOAD 1
  IRETURN
```
While the changes are not very impressive on that simple example, they become more important in more complex cases, notably with pattern matching. Examples can be found in the changed bytecode tests.

An added benefit is that `genLoadTo` knows when loading a value results in an unconditional control flow change (jump, return or throw). It can then avoid inserting any useless adaptation. This removes all the dead bytecode that the codegen used to generate as artifacts of its own compilation scheme. (It will still generate dead bytecode if the original source code/inlined code contains dead code.)

### CODE REVIEW

File: compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala [hunk 1]
```diff
@@ -578,54 +621,39 @@ trait BCodeBodyBuilder extends BCodeSkelBuilder {
...
+      markProgramPoint(loop)
+
       if isInfinite then
-        body match
-          case Labeled(bind, expr) if tpeTK(body) == UNIT =>
-            // this is the shape of tailrec methods
-            val loop = programPoint(bind.symbol)
-            markProgramPoint(loop)
-            genLoad(expr, UNIT)
-            bc goTo loop
```
Comment (sjrd) [inline compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala:600]: This optimization is now performed "by construction" of the generic `LoadDestination` infrastructure. :)

File: compiler/src/dotty/tools/backend/jvm/BCodeSkelBuilder.scala [hunk 2]
```diff
@@ -799,20 +811,16 @@ trait BCodeSkelBuilder extends BCodeHelpers {
 
         def emitNormalMethodBody(): Unit = {
           val veryFirstProgramPoint = currProgramPoint()
-          genLoad(trimmedRhs, returnType)
-
-          trimmedRhs match {
-            case (_: Return) | Block(_, (_: Return)) => ()
-            case (_: Apply) | Block(_, (_: Apply)) if trimmedRhs.symbol eq defn.throwMethod => ()
```
Comment (sjrd) [inline compiler/src/dotty/tools/backend/jvm/BCodeSkelBuilder.scala:806]: These two optimizations are also taken care of by the general `LoadDestination` infrastructure.

**Comment by sjrd** [comment]
Ping @lrytz ?

File: compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala [hunk 3]
```diff
@@ -578,54 +621,39 @@ trait BCodeBodyBuilder extends BCodeSkelBuilder {
...
-            markProgramPoint(loop)
             genCond(cond, success, failure, targetIfNoJump = success)
             markProgramPoint(success)
-            genLoad(body, UNIT)
-            bc goTo loop
+            genLoadTo(body, UNIT, LoadDestination.Jump(loop))
             markProgramPoint(failure)
         end match
-        UNIT
+        LoadDestination.FallThrough
```
Comment (lrytz) [inline compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala:656]: `genWhileDo` is the only method that returns a `LoadDestination`, and it constantly returns `FallThrough`, so maybe make the method `: Unit`?

**Review by lrytz — APPROVED** [review]
This is great, definitely worth backporting to Scala 2! Not that trivial, but LGTM, I couldn't spot any mistakes.

File: compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala (hunk 3 above)
Comment (lrytz) [reply compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala:656]: Right, I missed that. And the beginning of the `loop` destination then propagates down, which takes care of the tailrec handling.

Comment (sjrd) [reply compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala:656]: Yes, exactly. :)

**Pull request merged**